enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,229 @@
+> the loaded CNI and CSI plugins. The plugin developers may choose to add plugin-
+> specific health check scripts.
+
+Exiting the script with a non-zero status will have the boot declared as failed.
+Greenboot redirects all the script output to the system log, accessible via the
+`journalctl -u greenboot-healthcheck.service` command.
+
+|Validation                                        |Pass  |Fail  |
+|--------------------------------------------------|------|------|
+|Check microshift.service is enabled               |Next  |exit 0|
+|Wait for microshift.service to be active (!failed)|Next  |exit 1|
+|Wait for any pods to start                        |Next  |exit 1|
+|For each core namespace, wait for pods to be ready|Next  |exit 1|
+|For each core namespace, check pods not restarting|exit 0|exit 1|
+
+> If the system is not booted using the `ostree` file system, the health check
+> procedures still run, but no rollback would be possible in case of an upgrade
+> failure.
+
+The wait timeout of each validation starts from the base of 5 minutes and it is
+incremented by the base value after each boot with a failed verification (i.e.
+first failed boot 5 minutes wait, second failed boot 10 minutes wait, etc.).
+It should be possible to override the base wait timeout setting with the
+`MICROSHIFT_WAIT_TIMEOUT_SEC` environment variable in the `/etc/greenboot/greenboot.conf`
+configuration file alongside other [Greenboot Configuration](https://github.com/fedora-iot/greenboot#configuration)
+settings.
+
+### Risks and Mitigations
+
+Edge Devices often have relatively slow networks or occasional loss of network
+connectivity. The risk of declaring a `false positive` on the failed boot may
+be caused by these factors.
+
+If an update requires downloading new container images on a relatively slow
+network, it may take more than the default wait period to run this operation.
+The mitigation would be automatic doubling of the wait period on subsequent
+boots. If this is still not enough, Device Administrators should set the base
+wait period setting to a higher value to start with.
+
+If any external connectivity is required for upgrades to complete successfully,
+intermittent network outage may result in a rollback to the previous version.
+As a mitigation, Device Administrators should reinstall the upgrade when network
+is available for the Edge Devices.
+
+To further mitigate potential problems related to slow or unstable networks,
+Device Administrators may choose to bundle the upgraded container images in the
+`ostree` software upgrade. In this case, when the upgrade is installed, no
+more container images will need to be downloaded over the network.
+
+Some 3rd party workloads may become operational before the upgrade is declared
+valid and potentially create or update data on the device. If a rollback is
+performed subsequentially, there is a risk of data loss because the file system
```

Comment:
I think both. Any customizations made by user-specific workloads may be lost or become incompatible if the system is rolled back. 
Should I expand on this here?

Reply:
Yes. We are not going to be able to prevent workloads from starting. Consider the case that MicroShift reports it is OK and it is running 2 workloads. The first may start up, upgrade a database schema for a database on a PV, then report it is OK. The second may start up and report a failure. Greenboot would then trigger a reboot. If that second workload never recovers, then the first workload has to be able to rollback, but it's not MicroShift's fault.

Reply:
Added more information here.